fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
+            amount: _,
+            relay_eon: _,
+        } => Ok(HandleResponse::default()),
         HandleMsg::Pause { since_block: _ } => Ok(HandleResponse::default()),
-        HandleMsg::FreezeFunds {amount: _}  => Ok(HandleResponse::default()),
-        HandleMsg::UnFreezeFunds {amount: _}  => Ok(HandleResponse::default()),
+        HandleMsg::FreezeFunds { amount: _ } => Ok(HandleResponse::default()),
+        HandleMsg::UnFreezeFunds { amount: _ } => Ok(HandleResponse::default()),
         HandleMsg::SetCap { amount: _ } => Ok(HandleResponse::default()),
-        HandleMsg::SetLimits { swap_min: _, swap_max: _, swap_fee: _ } => Ok(HandleResponse::default()),
-        HandleMsg::GrantRole { role: _, address: _ } => Ok(HandleResponse::default()),
-        HandleMsg::RevokeRole { role: _, address: _ } => Ok(HandleResponse::default()),
-        HandleMsg::RenounceRole { role: _, address: _ } => Ok(HandleResponse::default()),
+        HandleMsg::SetLimits {
+            swap_min: _,
+            swap_max: _,
+            swap_fee: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::GrantRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::RevokeRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
+        HandleMsg::RenounceRole {
+            role: _,
+            address: _,
+        } => Ok(HandleResponse::default()),
     }
 }
 
 fn try_swap<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
-    _env: Env,
-    _amount: U256,
-    _destination: String,
+    env: &Env,
+    state: &State,
+    amount: Uint128,
+    destination: String,
 ) -> StdResult<HandleResponse> {
-   config(&mut deps.storage).update(|mut state| {
+    verify_not_paused(env, state)?;
+    verify_swap_amount(amount, state)?;
+
+    let swap_id = state.next_swap_id;
+    config(&mut deps.storage).update(|mut state| {
+        state.supply += amount;
         state.next_swap_id += 1;
         Ok(state)
     })?;
 
-    Ok(HandleResponse::default())
+    let log = vec![
+        log("action", "swap"),
+        log("destination", destination),
+        log("swap_id", swap_id),
+        log("amount", amount), // TOFIX(LR) how about fees?
+    ];
+
+    let r = HandleResponse {
+        messages: vec![],
+        log,
+        data: None,
+    };
+    Ok(r)
+}
+
+fn try_reverse_swap<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: &Env,
+    state: &State,
+    rid: u64,
+    to: HumanAddr,
+    from: String,
+    origin_tx_hash: Uint128,
+    amount: Uint128,
+    relay_eon: u64,
+) -> StdResult<HandleResponse> {
+    verify_tx_relay_eon(relay_eon, state)?;
+    only_relayer(env, state)?;
+
+    if amount > state.swap_fee {
+        // TOFIX(LR) when amount == fee, amount will still be consumed
+        let swap_fee = state.swap_fee;
+        let effective_amount = (amount - swap_fee)?;
```

Comment:
Does this detect undeflow, and fails if the undeflow occurs?
If not, we have to fix it.